docs/design/Path concatenation.md

```diff
@@ -0,0 +1,94 @@
+5. Repeat step 4 until no such quadruplets are left.
+6. Return the result of the above modifications to `path`.
+
+## `GetFullPath` specification
+
+`GetFullPath` is a function with two parameters: `appendPath`, which is a _relative path_, and `basePath`, which is an _absolute path_. The result of this funtion is an _absolute path_ or an error, whichever is determined by the an algorithm equivalent to the following:
+
+1. Count the number of _backtrack segments_ in `appendPath`.
+2. Delete as many _segments_ in `basePath`, starting from the end. If the operation cannot be performed due to `basePath` not having enough _segments_, return with error.
+3. _Normalize_ `basePath`.
+4. Return the result as the sequence of all _chars_ in the `basePath` followed by all _chars_ in the `appendPath`.
+
+## `GetRelativePath` specification
+
+`GetRelativePath` is a function with two parameters: `path`, which is an _absolute path_, and `basePath`, which is also an _absolutePath_. The result of this function is a _relative path_, determined by an algorithm equivalent to the following:
+
+1. If the _segments_ at the start of `path` and `basePath` are equal, delete them from both with the preceding separators.
+2. Repeat step 1 until the _segements_ at the start of each path are no longer equal.
+3. Count the number of segments in `basePath`.
+4. Prepend that number of _separators_ and _backtrack segments_, alternating between the two and starting with a _separator_, to `path`.
+5. Return the result of the above modifications to `path`.
+
+## Performance
+
+When working on any algorithm, it is critical that data used in benchmarks reflects the real world. This is because performance of modern CPUs hinges critically on how predictable the input data is. It is thus essential to tune the algorithm to branch in predictable patterns.
+
+As an illustration, this benchmark of a prototype implementation of `GetFullPath` illustrates the differences between predictable and unpredictable (data-dependent) control flow:
+![image](https://user-images.githubusercontent.com/62474226/104724925-6e9a5c00-5742-11eb-963b-d19859d5465a.png)
+
+Conceptually, the prototype copies all segments from the source paths into the result, looking for separators. Running it over data where there is a 50/50 chance the next `char` will be a separator will not be very performant. At the same time, fully branchless implementation would most certainly be slower in the `PureConcatNoSeparators` benchmark. As such it is critical to identify patterns that occur frequently in real-world paths and optimize for them, finding the "sweet spot" for branching.
+
+In practice, we are most interested in the distribution of segment lengths. Obviously, there are no real-world systems that use DF2 paths today, and so we have to use surrogates, like file system paths. Running the following Powershell script shows that the assumption of the average segment being 10 characters long is mostly correct.
```

Comment:
Note that System-to-System DF2 things are unlikely to use string paths anyways. Such use cases would likely prefer OID paths.

Also, I don't really understand what this section is about? what am I supposed to take from this?

Reply:
This sections goal is to define what a full path is, what a relative path is, and how they can be concatenated with eachother, which is required foreknowledge for some Str command specifications

Reply:
It is an artifact from when the formal specification was not yet in sight.
I can delete it.